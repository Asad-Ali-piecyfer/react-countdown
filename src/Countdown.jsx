import React from 'react';
import PropTypes from 'prop-types';

/**
 * Pads a given string or number with zeros.
 *
 * @param {any} value Value to zero-pad.
 * @param {number} [length=2] Amount of characters to pad.
 * @returns Left-padded string.
 */
export const zeroPad = (value, length = 2) => {
  const strValue = String(value);
  return strValue.length >= length ? strValue : ('0'.repeat(length) + strValue).slice(length * -1);
};

/**
 * Calculates the time difference between a given start date and the current date.
 *
 * @param {any} date Date or timestamp representation of the end date.
 * @param {any} [now=Date.now] Alternative function for returning the current date.
 * @param {number} [precision=0] The precision on a millisecond basis.
 * @param {boolean} [controlled=false] Defines whether the calculated value is already provided as the time difference or not.
 * @returns Object that includes details about the time difference.
 */
export const getTimeDifference = (date, now = Date.now, precision = 0, controlled = false) => {
  const startDate = typeof date === 'string' ? new Date(date) : date;
  const total = parseInt(
    (Math.max(0, controlled ? startDate : startDate - now()) / 1000).toFixed(precision) * 1000,
    10
  );

  const seconds = total / 1000;

  return {
    total,
    days: Math.floor(seconds / (3600 * 24)),
    hours: Math.floor(seconds / 3600 % 24),
    minutes: Math.floor(seconds / 60 % 60),
    seconds: Math.floor(seconds % 60),
    milliseconds: Number((seconds % 1 * 1000).toFixed()),
  };
};

export default class Countdown extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      ...getTimeDifference(props.date, props.now, props.precision, this.props.controlled),
    };
  }

  componentDidMount() {
    if (!this.props.controlled) {
      this.interval = setInterval(this.tick, this.props.intervalDelay);
    }
  }

  componentWillReceiveProps(nextProps) {
    this.setDeltaState(
      getTimeDifference(nextProps.date, nextProps.now, nextProps.precision, nextProps.controlled)
    );
  }

  componentWillUnmount() {
    this.clearInterval();
  }

  setDeltaState(delta) {
    if (this.state.total > 0 && delta.total <= 0) {
      this.clearInterval();

      if (this.props.onComplete) {
        this.props.onComplete(delta);
      }
    }

    this.setState({ ...delta });
  }

  getFormattedDelta() {
    let { days, hours } = this.state;
    const { minutes, seconds } = this.state;
    const { daysInHours, zeroPadLength } = this.props;

    if (daysInHours) {
      hours += days * 24;
      days = null;
    } else {
      days = zeroPad(days, zeroPadLength);
    }

    return {
      days,
      hours: zeroPad(hours, zeroPadLength),
      minutes: zeroPad(minutes, Math.min(2, zeroPadLength)),
      seconds: zeroPad(seconds, Math.min(2, zeroPadLength)),
    };
  }

  clearInterval() {
    clearInterval(this.interval);
    delete this.interval;
  }

  tick = () => {
    const delta = getTimeDifference(
      this.props.date,
      this.props.now,
      this.props.precision,
      this.props.controlled
    );
    this.setDeltaState({
      ...delta,
    });

    if (this.props.onTick && delta.total > 0) {
      this.props.onTick(delta);
    }
  };

  render() {
    if (this.props.renderer) {
      return this.props.renderer({ ...this.props, ...this.state });
    }

    const { days, hours, minutes, seconds } = this.getFormattedDelta();
    return <span>{days}{days ? ':' : ''}{hours}:{minutes}:{seconds}</span>;
  }
}

Countdown.propTypes = {
  date: PropTypes.oneOfType([PropTypes.instanceOf(Date), PropTypes.string, PropTypes.number])
    .isRequired, // eslint-disable-line react/no-unused-prop-types
  daysInHours: PropTypes.bool,
  zeroPadLength: PropTypes.number,
  controlled: PropTypes.bool,
  intervalDelay: PropTypes.number,
  precision: PropTypes.number,
  renderer: PropTypes.func,
  now: PropTypes.func, // eslint-disable-line react/no-unused-prop-types
  onTick: PropTypes.func,
  onComplete: PropTypes.func,
};

Countdown.defaultProps = {
  daysInHours: false,
  zeroPadLength: 2,
  controlled: false,
  intervalDelay: 1000,
  precision: 0,
};
