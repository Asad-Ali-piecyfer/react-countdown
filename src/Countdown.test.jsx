import React from 'react';
import { mount, shallow } from 'enzyme';
import Countdown, { leftPad, getTimeDifference } from './Countdown';

const timeDiff = 90110456;
const now = jest.fn(() => 1482363367071);
Date.now = now;

describe('<Countdown />', () => {
  jest.useFakeTimers();

  it('compares snapshot of countdown with custom renderer', () => {
    const wrapper = shallow(
      <Countdown
        date={Date.now() + timeDiff}
        renderer={props => <div>{props.days}{props.hours}{props.minutes}{props.seconds}</div>}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('compares snapshot of countdown with daysInHours => true', () => {
    const wrapper = shallow(<Countdown date={Date.now() + timeDiff} daysInHours />);
    expect(wrapper).toMatchSnapshot();
  });

  let wrapper = null;
  let wrapperDate = null;

  beforeEach(() => {
    const date = Date.now() + 10000;
    const root = document.createElement('div');
    wrapperDate = date;
    wrapper = mount(<Countdown date={date} />, { attachTo: root });
  });

  it('should trigger onTick and onComplete callbacks', () => {
    const onTick = jest.fn(stats => {
      expect(stats).toEqual(getTimeDifference(wrapperDate, Date.now));
    });

    const onComplete = jest.fn(stats => {
      expect(stats.total).toEqual(0);
    });

    wrapper.setProps({ onTick, onComplete });
    expect(onTick).not.toBeCalled();

    // Forward 6s in time
    Date.now = jest.fn(() => wrapperDate - 6000);
    jest.runTimersToTime(6000);
    expect(onTick.mock.calls.length).toBe(6);
    expect(wrapper).toMatchSnapshot();

    // Forward 3 more seconds
    Date.now = jest.fn(() => wrapperDate - 1000);
    jest.runTimersToTime(3000);
    expect(onTick.mock.calls.length).toBe(9);
    expect(wrapper.state().seconds).toBe(1);

    // The End: onComplete callback gets triggered instead of onTick
    Date.now = jest.fn(() => wrapperDate);
    jest.runTimersToTime(1000);
    expect(onTick.mock.calls.length).toBe(9);
    expect(onComplete.mock.calls.length).toBe(1);
  });

  it('should run through the controlled component by updating the date prop', () => {
    const root = document.createElement('div');
    wrapper = mount(<Countdown date={1000} controlled />, { attachTo: root });
    expect(wrapper.instance().interval).toBeUndefined();

    wrapper.setProps({ date: 0 });
    expect(wrapper.state().total).toBe(0);
  });

  afterEach(() => {
    wrapper.detach();
  });
});

describe('leftPad', () => {
  it('should add two 0s in front of 2', () => {
    expect(leftPad(2, 3)).toBe('002');
  });

  it('should add one 0 in front of "ab"', () => {
    expect(leftPad('ab', 3)).toBe('0ab');
  });

  it('should add three 0s', () => {
    expect(leftPad('', 3)).toBe('000');
  });

  it('should not left-pad 1 when length is not defined or 0', () => {
    expect(leftPad(1)).toBe('1');
    expect(leftPad(1, 0)).toBe('1');
  });

  it('should not left-pad 123 when passing 3 as length', () => {
    expect(leftPad(123, 3)).toBe('123');
  });
});

describe('getTimeDifference', () => {
  const stats = {
    total: 0,
    days: 0,
    hours: 0,
    minutes: 0,
    seconds: 0,
    milliseconds: 0,
  };

  it('should return a time difference of 0 seconds', () => {
    expect(getTimeDifference(Date.now())).toEqual(stats);
  });

  it('should return a time difference of 0 seconds if values for start and current date are the same', () => {
    expect(getTimeDifference(Date.now(), Date.now)).toEqual(stats);
    expect(getTimeDifference(Date.now() + 10, () => Date.now() + 10)).toEqual(stats);
  });

  it('should calculate the time difference', () => {
    expect(getTimeDifference(Date.now() + timeDiff)).toEqual({
      total: timeDiff,
      days: 1,
      hours: 1,
      minutes: 1,
      seconds: 50,
      milliseconds: 456,
    });
  });

  it('should calculate the time difference by passing a date string', () => {
    Date.now = jest.fn(() => new Date('Thu Dec 22 2016 00:36:07').getTime());
    expect(getTimeDifference('Thu Dec 23 2017 01:38:10:456')).toEqual({
      total: 31626123456,
      days: 366,
      hours: 1,
      minutes: 2,
      seconds: 3,
      milliseconds: 456,
    });
  });

  it('should calculate the time difference when controlled is true', () => {
    const total = 91120003;
    expect(getTimeDifference(total, undefined, true)).toEqual({
      total,
      days: 1,
      hours: 1,
      minutes: 18,
      seconds: 40,
      milliseconds: 3,
    });
  });
});
