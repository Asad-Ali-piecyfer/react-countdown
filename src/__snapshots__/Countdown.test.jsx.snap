// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Countdown /> compares snapshot of countdown with custom renderer 1`] = `
<div>
  1
  1
  1
  50
</div>
`;

exports[`<Countdown /> compares snapshot of countdown with daysInHours => true 1`] = `
<span>
  25
  :
  01
  :
  50
</span>
`;

exports[`<Countdown /> should trigger onTick and onComplete callbacks 1`] = `
<Countdown
  controlled={false}
  date={1482363377071}
  daysInHours={false}
  intervalDelay={1000}
  onComplete={[Function]}
  onTick={[Function]}
  precision={0}
  zeroPadLength={2}
>
  <span>
    00
    :
    00
    :
    00
    :
    06
  </span>
</Countdown>
`;
